Optimize alerts detection

davidor opened the pull request:
Currently, the code is loading all the usage limits of a plan and the current usages for those limits on every report job.

This PR introduces a new class (`UsageChecked`) that allows us to do just once and then, no longer do it for the same application until:
1) An alert expires (24h passed since it was triggered) and can be raised again.
2) A new alert bin is enabled for the service.

With this, on every report job we just need to check the limits and usages of the metrics increased in the job. This should have a noticeable impact for users with many usage limits defined.

unleashed commented on lib/3scale/backend/alerts.rb:
```diff
@@ -45,6 +50,69 @@ def key_current_id
+          ttl = ALERT_BINS.map do |bin|
+            ttl = storage.ttl(key_already_notified(service_id, app_id, bin))
+
+            # Redis returns -2 is key does not exist, and -1 if it exists without
```
typo `s/is/if`

davidor replied:
fixed :+1:

unleashed commented on test/unit/alerts_test.rb:
```diff
@@ -46,4 +49,69 @@ def setup
+
+    assert_false Alerts::UsagesChecked.need_to_check_all?(@service_id, @application_id)
+
+    # The 80% bin has a lower TTL. We cannot guarantee the exact TTL because
+    # some time might pass between the moment we create the key and check its
+    # TTL, but we know it needs to be <= min_ttl and close to it.
```
So maybe just ensure the offset is something bigger than a time difference since before you create the key and right before you check its TTL to reduce the odds of spurious failures?

davidor replied:
I think that's difficult to get exactly right, that's why I used a range instead.

unleashed commented on test/unit/alerts_test.rb:
```diff
@@ -46,4 +49,69 @@ def setup
+
+    assert_false Alerts::UsagesChecked.need_to_check_all?(@service_id, @application_id)
+
+    # We can't guarantee that the TTL will be exactly one day because some
+    # seconds might pass between the moment we create the key and check the TTL,
+    # but it should be pretty close.
```
Ditto.

davidor replied:
https://github.com/3scale/apisonator/pull/287#discussion_r650953996

unleashed commented on lib/3scale/backend/stats/aggregator.rb:
```diff
@@ -141,14 +141,23 @@ def update_alerts(transactions)
               begin
-                max_utilization = Utilization.max_in_all_metrics(service_id, app_id)
+                max_utilization = if Alerts::UsagesChecked.need_to_check_all?(service_id, app_id)
+                                    max = Utilization.max_in_all_metrics(service_id, app_id)
+                                    Alerts::UsagesChecked.mark_all_checked(service_id, app_id)
+                                    max
```
Great opportunity to `tap` a Ruby idiom:

```ruby
Utilization.max_in_all_metrics(service_id, app_id).tap do
  Alerts::UsagesChecked.mark_all_checked service_id, app_id
end
```

davidor replied:
Changed :+1:

unleashed reviewed (approved):
LGTM but check comments!

unleashed commented on lib/3scale/backend/stats/aggregator.rb:
```diff
@@ -141,14 +141,23 @@ def update_alerts(transactions)
+                                    # metrics_ids here includes the metrics
+                                    # explicitly reported plus their parents in
+                                    # the hierarchy.
+                                    metric_ids = txs.map { |tx| tx.usage.keys }.flatten.uniq
```
Hm, didn't a test fail because of lacking `flatten` here?

davidor replied:
No, we need more tests in this part. I'm working on it.

davidor commented:
@unleashed I added a new commit with the tests that were missing. It's the last commit.
I also added a `Memoizer.clear` call here: https://github.com/3scale/apisonator/pull/287/files#diff-b176f67b39cc7bd8b50fe54aa2a4de9cf404809e691aab4a2199d6aa8dc10011R84

unleashed commented on lib/3scale/backend/alerts.rb:
```diff
@@ -45,43 +50,82 @@ def key_current_id
-            end
-          end
+          storage.setex(key_usage_already_checked(service_id, app_id), ttl, '1'.freeze)
+          Memoizer.clear(Memoizer.build_key(self, :need_to_check_all?, service_id, app_id))
```
Rather than clearing you could just set the new memoized value for `:need_to_check_all?` (would avoid one command per memorization cycle, so no big deal).

davidor replied:
That would increase complexity a bit in my opinion.
I don't think there's a clear winner in this case, bud I'd rather leave it as it is now.

unleashed commented:
bors delegate+

davidor commented:
bors r+
